### PR TITLE
Prevent job workers from waking up on new job while another one is running

### PR DESCRIPTION
**Pull request by DrDet**
The main idea is to enable `EPOLLONESHOT` flag for fd we read new jobs from and rearm it only when job is already served.

relates to `KPHP-1353`

### CODE REVIEW

File: server/job-workers/job-worker-server.h [hunk 1]
```diff
@@ -47,6 +49,8 @@ class JobWorkerServer : vk::not_copyable {
   bool reply_is_expected() const noexcept;
 
 private:
+  static const int EPOLL_FLAGS;
```
Comment (drdzyk) [inline server/job-workers/job-worker-server.h:52]: constexpr static? Maybe place this var directly in source file?

File: server/job-workers/job-worker-server.cpp [hunk 2]
```diff
@@ -129,10 +133,13 @@ int JobWorkerServer::job_parse_execute(connection *c) {
     assert(errno == EWOULDBLOCK);
     // another job worker has already taken the job (all job workers are readers for this fd)
     // or there are no more jobs in pipe
+    tvkprintf(job_workers, 3, "No jobs in pipe after wakeup\n");
     ++vk::singleton<SharedMemoryManager>::get().get_stats().job_worker_skip_job_due_steal;
+    rearm_read_job_fd(); // because > 1 workers can wake up on single job
```
Comment (drdzyk) [inline server/job-workers/job-worker-server.cpp:138]: I think it worth to wrap rearm_read_job_fd() call in some RAII class. Currently it is possible forget to made such call and fd will hang.

File: net/net-reactor.cpp [hunk 3]
```diff
@@ -323,7 +326,7 @@ int net_reactor_insert(net_reactor_ctx_t *ctx, int fd, int flags) {
     return 0;
   }
 
-  flags &= EVT_NEW | EVT_NOHUP | EVT_LEVEL | EVT_RWX;
+  flags &= EVT_NEW | EVT_NOHUP | EVT_LEVEL | EVT_RWX | EPOLLONESHOT;
```
Comment (drdzyk) [inline net/net-reactor.cpp:329]: Maybe add own flag for EPOLLONESHOT as it done for e.g. EVT_READ <-> EPOLLIN?

**Review by drdzyk — CHANGES_REQUESTED** [review]

File: server/job-workers/job-worker-server.h (hunk 1 above)
Comment (DrDet) [reply server/job-workers/job-worker-server.h]: ok, done

File: net/net-reactor.cpp (hunk 3 above)
Comment (DrDet) [reply net/net-reactor.cpp:329]: I think it's strange that we have own flags which map to library flags (EVT_READ <-> EPOLLIN and so on). So, I think it's better not to create new own flags. For example, EPOLLEXCLUSIVE doesn't have other own flag too

File: server/job-workers/job-worker-server.cpp (hunk 2 above)
Comment (DrDet) [reply server/job-workers/job-worker-server.cpp]: done

File: net/net-reactor.cpp (hunk 3 above)
Comment (drdzyk) [reply net/net-reactor.cpp:329]: Sure

**Review by drdzyk — DISMISSED** [review]

File: server/job-workers/job-worker-server.h [hunk 4]
```diff
@@ -36,6 +36,8 @@ class JobWorkerServer : vk::not_copyable {
 
   void init();
 
+  void rearm_read_job_fd();
```
Comment (unserialize) [inline server/job-workers/job-worker-server.h:39]: noexcept?

File: server/job-workers/job-worker-server.cpp [hunk 5]
```diff
@@ -188,6 +203,12 @@ void JobWorkerServer::init() {
   job_reader = PipeJobReader{read_job_fd};
 }
 
+
+void JobWorkerServer::rearm_read_job_fd() {
+  // We need to rearm fd because we use EPOLLONESHOT
+  epoll_insert(read_job_fd, EPOLL_FLAGS);
+}
+
```
Comment (unserialize) [inline server/job-workers/job-worker-server.cpp:211]: Seems like an extra newline

Comment (DrDet) [reply server/job-workers/job-worker-server.cpp:206]: fixed whitespaces

File: server/job-workers/job-worker-server.h (hunk 4 above)
Comment (DrDet) [reply server/job-workers/job-worker-server.h]: added for all of the JobWorkerServer functions

**Review by unserialize — APPROVED** [review]

**Pull request merged**